syntax = "proto3";

package datastore;

import "google/protobuf/timestamp.proto";

option go_package = "./datastore";

// Notes:
// - A _time series_ is a context defined by a set of metadata (defined in TSMetadata below) that
//   usually does not vary with observaion (time).
// - An _observation_ consists of a set of of metadata (defined in ObsMetadata below) that usually
//   varies with observation (time). Note that for simplicity the observation value itself (such as
//   air temperature value 12.7) is also considered a metadata field (although strictly speaking
//   this is the only field that is just data, not metadata).
// - There is a 1:N relationship between time series and observations:
//   * A given time series (= unique combination of time series metadata fields) is associated with
//     one or more observations.
//   * A given observation (= unique combination of observation metadata fields) is associated with
//     exactly one time series.

service Datastore {
  // insert observations into the storage (or update existing ones)
  rpc PutObservations(PutObsRequest) returns (PutObsResponse);

  // retrieve observations from the storage
  rpc GetObservations(GetObsRequest) returns (GetObsResponse);

  // get unique combinations of a set of time series attributes
  rpc GetTSAttrGroups(GetTSAGRequest) returns (GetTSAGResponse);

  // get temporal and spatial extents of current storage contents
  rpc GetExtents(GetExtentsRequest) returns (GetExtentsResponse);
}

//---------------------------------------------------------------------------

message Point { // horizontal position
  double lat = 1; // latitude degrees in range [-90, 90]
  double lon = 2; // longitude degrees in range [-180, 180]
}

message Polygon { // horizontal area; three or more points
  repeated Point points = 1;
}

message BoundingBox {
  double left = 1;
  double bottom = 2;
  double right = 3;
  double top = 4;
}

message TimeInterval {
  google.protobuf.Timestamp start = 1;
  google.protobuf.Timestamp end = 2;
}

message Link {
  string href = 1;
  string rel = 2;
  string type = 3;
  string hreflang = 4;
  string title = 5;
}

message TSMetadata {
  string version = 1;
  string type = 2;
  string title = 3;
  string summary = 4;
  string keywords = 5;
  string keywords_vocabulary = 6;
  string license = 7;
  string conventions = 8;
  string naming_authority = 9;
  string creator_type = 10;
  string creator_name = 11;
  string creator_email = 12;
  string creator_url = 13;
  string institution = 14;
  string project = 15;
  string source = 16;
  string platform = 17;
  string platform_vocabulary = 18;
  string standard_name = 19;
  string unit = 20;
  string instrument = 21;
  string instrument_vocabulary = 22;
  repeated Link links = 23;
}

message ObsMetadata {
  string id = 1;
  oneof geometry {
    Point geo_point = 2;
    Polygon geo_polygon = 3;
  }
  google.protobuf.Timestamp pubtime = 4;
  string data_id = 5;
  string history = 6;
  string metadata_id = 7;
  oneof obstime {
    google.protobuf.Timestamp obstime_instant = 8;
    //TimeInterval obstime_interval = 9; -- unsupported for now
  }
  string processing_level = 10;
  string value = 11;
}

//---------------------------------------------------------------------------

message Metadata1 { // denormalized (more redundancy)
  TSMetadata ts_mdata = 1;
  ObsMetadata obs_mdata = 2;
}

message Metadata2 { // normalized  (less redundancy)
  TSMetadata ts_mdata = 1;
  repeated ObsMetadata obs_mdata = 2;
}

//---------------------------------------------------------------------------

message PutObsRequest {
  repeated Metadata1 observations = 1;
}

message PutObsResponse {
  int32 status = 1;
  string error = 2; // any error description (empty on success)
}

//---------------------------------------------------------------------------

message GetObsRequest {
  TimeInterval interval = 1; // only return observations in this time range
  Polygon inside = 2; // if specified, only return observations in this area
  repeated string platforms = 3; // if specified, only return observations matching any of these platform patterns
  repeated string standard_names = 4; // if specified, only return observations matching any of these standard names
  repeated string instruments = 5; // if specified, only return observations matching any of these instruments
  repeated string processing_levels = 6; // if specified, only return observations matching any of these processing levels
  // TODO: add search filters for other metadata
}

message GetObsResponse {
  int32 status = 1;
  string error = 2; // any error description (empty on success)
  repeated Metadata2 observations = 3;
}

//---------------------------------------------------------------------------

message TSMdataGroup {
  TSMetadata combo = 1;
  repeated TSMetadata instances = 2;
}

message GetTSAGRequest {
  repeated string attrs = 1; // one or more TSMetadata field names
  bool include_instances = 2; // whether to include all matching time series instances in response
}

message GetTSAGResponse {
  repeated TSMdataGroup groups = 1; // zero or more groups of unique combinations of
  // GetTSAGRequest.attrs currently existing in the storage.
  //
  // - The combo of each group always contains the GetTSACRequest.attrs in question, with all other
  //   attributes set to default values (typically empty strings).
  //
  // - If GetTSAGRequest.include_instances is true, the instances array of each group contains
  //   all time series instances that match GetTSACRequest.attrs (note that GetTSAGRequest.attrs
  //   will be repeated in each item).
  //
  // - If GetTSAGRequest.include_instances is false (the default), the instances arrays are omitted.
}

//---------------------------------------------------------------------------

message GetExtentsRequest {
  // currently no args
}

message GetExtentsResponse {
  int32 status = 1;
  string error = 2; // any error description (empty on success)
  TimeInterval temporal_extent = 3;
  BoundingBox spatial_extent = 4;
}
